### PR TITLE
expression: implement vectorized evaluation for `builtinCoalesceXXSig`

### PR DESCRIPTION
**Pull request by gtygo**
### What problem does this PR solve? 

use go generate to implement vectorized evaluation for Coalesce
### What is changed and how it works?
```
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceIntSig-VecBuiltinFunc-12              	  189350	      6218 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceIntSig-NonVecBuiltinFunc-12           	   59558	     19551 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceRealSig-VecBuiltinFunc-12             	  190057	      6094 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceRealSig-NonVecBuiltinFunc-12          	   61264	     18935 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceDecimalSig-VecBuiltinFunc-12          	  106948	     10088 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceDecimalSig-NonVecBuiltinFunc-12       	   52993	     22329 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceStringSig-VecBuiltinFunc-12           	   81129	     14239 ns/op	      32 B/op	       1 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceStringSig-NonVecBuiltinFunc-12        	   40249	     29148 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceTimeSig-VecBuiltinFunc-12             	  131838	      8231 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceTimeSig-NonVecBuiltinFunc-12          	   31866	     37015 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceDurationSig-VecBuiltinFunc-12         	  183813	      6232 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceDurationSig-NonVecBuiltinFunc-12      	   62823	     18709 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceJSONSig-VecBuiltinFunc-12             	   81266	     14039 ns/op	      32 B/op	       1 allocs/op
BenchmarkVectorizedGeneratedBuiltinCompareFunc/builtinCoalesceJSONSig-NonVecBuiltinFunc-12          	   45518	     25790 ns/op	       0 B/op	       0 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by SunRunAway** [comment]
Hi @gtygo ,
Thanks for your pull request, you should commit both the generated code and generator.
And since #12875 is merged, you may need to combine your generator/compare_vec.go into his, then run `go generate` again, so that the conflicts in the generated files can be resolved.

File: expression/generator/compare_vec.go [hunk 1]
```diff
@@ -170,6 +170,92 @@ func (b *builtin{{ .compare.CompareName }}{{ .type.TypeName }}Sig) vectorized()
...
+	}
+	return nil
+}
+{{ else }}
+func (b *builtin{{ .compare.CompareName }}{{ .type.TypeName }}Sig) vecEval{{ .type.TypeName }}(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+
+	result.Reserve{{ .type.TypeNameInColumn }}(n)
+
+	buf1, err := b.bufAllocator.get(types.ET{{ .type.ETName }}, n)
```
Comment (qw4990) [inline expression/generator/compare_vec.go]: In most cases, `Coalesce` doesn't have many arguments, so how about allocating `len(b.args)` buffer columns at once.
Then your code would look like:
```
bufs := allocateBufs()
for i := 0; i < len(b.args); i++ {
    b.args.VecEvalString(bufs[i])
}

result.ReserveString(n)
for i := 0; i < n; i++ {
    for j := 0; j < len(b.args); j++ {
        .......
    }
}
```

File: expression/generator/compare_vec.go [hunk 2]
```diff
@@ -243,14 +344,27 @@ func generateDotGo(fileName string, compares []CompareContext, types []TypeConte
...
 				if err != nil {
 					return err
 				}
+			} else if compareCtx.CompareName == "Coalesce" {
+
+				err := builtinCoalesceCompareVecTpl.Execute(w, ctx)
+				if err != nil {
+					return err
+				}
+				if typesIndex == 0 {
```
[the remark this answers is not in the document]
Comment (gtygo) [reply expression/generator/compare_vec.go]: Yes, my initial idea is the same, but LTInt / GTInt /. Has been implemented in builtin_compare_vec.go. should i need to delete them and regenerate them with Go generate?

File: expression/generator/compare_vec.go (hunk 1 above)
Comment (qw4990) [reply expression/generator/compare_vec.go]: And please update the benchmark result in your PR's description after updating. @gtygo

**Comment by gtygo** [comment]
hi @SunRunAway , I  regenerated (LT/LE/GT/GE/EQ/NE/NullEQ Int)  by go generate.

File: expression/generator/compare_vec.go [hunk 3]
```diff
@@ -251,6 +341,13 @@ func generateDotGo(fileName string, compares []CompareContext, types []TypeConte
 				if err != nil {
 					return err
 				}
+			} else if compareCtx.CompareName == "Coalesce" {
```
Comment (SunRunAway) [inline expression/generator/compare_vec.go]: For `Coalesce`, this is right.
But for `NullEQ` and `ET/GE...`, we want to generate all types except int type and write int type by hands, so here we need to continue if we meet int type.

File: expression/generator/compare_vec.go [hunk 4]
```diff
@@ -273,6 +370,20 @@ func generateTestDotGo(fileName string, compares []CompareContext, types []TypeC
 	w.WriteString(builtinCompareVecTestHeader)
 
 	for _, compareCtx := range compares {
+		if compareCtx.CompareName == "Coalesce" {
```
Comment (SunRunAway) [inline expression/generator/compare_vec.go]: ditto

File: expression/generator/compare_vec.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (SunRunAway) [reply expression/generator/compare_vec.go]: Not to delete them and regenerate, but continue if we meet int type in LT/GT and NullEQ.

File: expression/generator/compare_vec.go (hunk 3 above)
Comment (gtygo) [reply expression/generator/compare_vec.go]: understood

File: expression/generator/compare_vec.go [hunk 5]
```diff
@@ -209,6 +279,20 @@ func BenchmarkVectorizedGeneratedBuiltinCompareFunc(b *testing.B) {
...
+var builtinCoalesceCompareVecTestFunc = template.Must(template.New("").Parse(`
+{
+retEvalType: types.ET{{ .ETName }},
+childrenTypes: []types.EvalType{types.ET{{ .ETName }}, types.ET{{ .ETName }}, types.ET{{ .ETName }}},
+geners: []dataGenerator{
+gener{defaultGener{eType: types.ET{{ .ETName }}, nullRation: 0.2}},
+gener{defaultGener{eType: types.ET{{ .ETName }}, nullRation: 0.2}},
+gener{defaultGener{eType: types.ET{{ .ETName }}, nullRation: 0.2}},
+},
+},
```
Comment (SunRunAway) [inline expression/generator/compare_vec.go]: nit: It's better to format these code.

File: expression/generator/compare_vec.go [hunk 6]
```diff
@@ -273,11 +372,28 @@ func generateTestDotGo(fileName string, compares []CompareContext, types []TypeC
...
+			}
+			w.WriteString(builtinCompareVecTestFuncTail)
+			continue
+		}
 		err := builtinCompareVecTestFuncHeader.Execute(w, compareCtx)
 		if err != nil {
 			return err
 		}
 		for _, typeCtx := range types {
+			if typeCtx.TypeName == "Int" {
```
Comment (SunRunAway) [inline expression/generator/compare_vec.go]:
```suggestion
			if typeCtx.TypeName == TypeInt.TypeName {
```

File: expression/generator/compare_vec.go [hunk 7]
```diff
@@ -247,11 +333,24 @@ func generateDotGo(fileName string, compares []CompareContext, types []TypeConte
 			ctx["compare"] = compareCtx
 			ctx["type"] = typeCtx
 			if compareCtx.CompareName == "NullEQ" {
+				if typeCtx.TypeName == "Int" {
```
Comment (SunRunAway) [inline expression/generator/compare_vec.go]:
```suggestion
				if typeCtx.TypeName == TypeInt.TypeName {
```

**Pull request merged**